### ISSUE TITLE
Option to view logs for each container

### ISSUE THREAD

**Issue by RamiKassouf** [rancher-sandbox/rancher-desktop#6386]
### Problem Description

Rancher desktop GUI is rendered useless by the fact that you can have a vscode extention that shows you everything.

There is no way to view the logs or get access to them unless you go to the cmd and run docker logs on it

### Proposed Solution

After right-clicking the 3 dots on the right of a container, one should have a View logs option

**Comment by jandubois** [comment]
Hi @RamiKassouf, I completely I agree that the container dashboard should allow browsing of the container logs (and even, when possible, to run a shell inside the container). Unfortunately the developer who started the container dashboard has moved to a different project, and I don't see anybody else picking this up any time soon.

For now your best bet would be to check out the "Logs Explorer" extension:

![CleanShot 2024-01-30 at 12 47 03@2x](https://github.com/rancher-sandbox/rancher-desktop/assets/78947/6e9faa78-506f-45c6-adaf-3a3f7c9db28b)

**Comment by thoom** [comment]
I've tried this extension and while it works to give you container logs, it truncates the container name in the left nav (which doesn't happen on the Docker Desktop version).

**Comment by jandubois** [comment]
> it truncates the container name in the left nav

Yeah, I noticed that too; I've filed a separate issue for that (#6453).

**Comment by mark1566** [comment]
The logs explorer would be better if it had a simple console output as per docker when viewed in IntelliJ. I  find this extension quite poor

**Comment by jandubois** [comment]
> the extension is quite poor in this regard

I agree. The extension however is provided by Docker and not by Rancher Desktop, so there is not much we can do about it (except maybe check to apply an upstream update).

> and I would also favour the feature proposed of Logs (and Terminal ?) being available per Container via the ellipsis menu.

And I agree with that as well. But as I mentioned before, we have a lack of experienced UI developers available to us right now, so we are limited in what we can do. Contributions welcome... 😄

**Comment by jandubois** [comment]
@bartvanandel Yes, it would be possible, if it worked with Rancher Desktop. I tried it (with moby) and it didn't display any running containers for me. Did it work for you?

In case you are not aware, you can always install non-listed extensions from the CLI, e.g.

```console
❯ rdctl extension install edwinat/loglens:2.0.1
Installing image edwinat/loglens:2.0.1: Created
```

I would recommend to always start testing with `moby` because Docker Desktop doesn't support `containerd`, so extension will not have been tested with it before.

But even with `moby` it is possible that extensions don't work properly because we may not have implemented a specific API that they rely upon.

**Comment by bartvanandel** [comment]
> In case you are not aware, you can always install non-listed extensions from the CLI

Did not know that, thanks. I'm currently trying (with `moby`, my regular config anyway), and indeed, no logs show up there.

I do have Logs Explorer still installed, and it became apparent that LogLens tries to access `/var/run/docker.sock` inside the container, which I confirmed doesn't exist. It is trying to mount `/var/run/docker.sock.raw` from the host using `docker-compose.yml`, as per [the documentation](https://docs.docker.com/extensions/extensions-sdk/guides/use-docker-socket-from-backend/), however this file doesn't exist (not even with Administrative Access enabled, which I'd rather avoid if possible to be honest, but this may be a necessary evil).

I don't know how Docker Desktop handles this, maybe it intercepts this as a special case?

**Comment by bartvanandel** [comment]
I've just compared the install details (i.e., files on host disk, under `$HOME/Library/rancher-desktop/extensions/...` on my Mac) between Logs Explorer and LogLens a bit.

For Logs Explorer, pretty much only the js app can be found here.
For LogLens however, it looks like there's a copy of the entire container file system here, including the `docker-compose.yaml` file (and a generated? `compose.yaml` file). I briefly tried to modify the compose file and re-run it, but its contents are overwritten when I restart Rancher Desktop.

Unfortunately I presently don't have the time to investigate any further. Maybe these findings can be helpful to someone else though, so I thought I'd share them here.

**Comment by bartvanandel** [comment]
Are you sure that `docker.sock.raw` is _supposed_ to be just a symlink to `docker.sock`? The API doc page I linked earlier warns about possible "permission issues in containers" so I'm guessing it's something slightly different.

Could it be that it's supposed to use `SOCK_RAW` as opposed to `SOCK_STREAM` or whatever the default is for regular sockets (see e.g. [socket(2)](https://www.man7.org/linux/man-pages/man2/socket.2.html])? From what I gather this provides access to whole packets instread of just the payload. But that's just from reading up on it a bit, I don't have much experience with it myself.

**Comment by mook-as** [comment]
Per the docs, `docker.sock.raw` is supposed to be what is visible in the VM, and `docker.sock` is the socket as seen from the host (i.e. goes through fun things like path translation, presumably).

So `rdctl shell` gets you into the VM, so it works fine.  If anything, our issue is more that we don't have a non-`raw` socket in the VM?

**Comment by mook-as** [comment]
Poked at LogLens a bit; it seems like there's an issue with their mongodb query? Filed https://github.com/edwin-abraham-thomas/LogLens/pull/17 (at least, that seems to solve the issue locally for me).

![Image](https://github.com/user-attachments/assets/81a4080a-ec44-44f2-8374-3ca51e6aad58)

I was not good enough with React to fix an issue where I can paginate past the last entry.

**Comment by binginsin** [comment]
So since I couldn't find any extensions that did this in a convenient fashion, I came up with https://github.com/binginsin/LogFollower
Also available in dockerhub as binginsin/LogFollower:1.0.0
It does it similarly to docker desktop, where it essentially just follows the container output. Colored logs supported.

![Image](https://github.com/user-attachments/assets/b2bdb327-fa2a-4f5f-81dc-741ae2d03bc6)

**Comment by jandubois** [comment]
> Also available in dockerhub as binginsin/LogFollower:1.0.0

It is actually called `binginsin/log-follower:1.0.0`

It looks fine in light mode:

![Image](https://github.com/user-attachments/assets/4e2a82d7-5442-43cd-93b0-1f00cb5d8523)

But in dark mode the box lines are barely visible:

![Image](https://github.com/user-attachments/assets/16979bc4-e5b0-44b6-b983-fd714c67fa51)

**Comment by DainiusCerniauskasAngstrom** [comment]
> > Also available in dockerhub as binginsin/LogFollower:1.0.0
> 
> It is actually called `binginsin/log-follower:1.0.0`
> 
> It looks fine in light mode:
> 
> ![Image](https://github.com/user-attachments/assets/4e2a82d7-5442-43cd-93b0-1f00cb5d8523)
> 
> But in dark mode the box lines are barely visible:
> 
> ![Image](https://github.com/user-attachments/assets/16979bc4-e5b0-44b6-b983-fd714c67fa51)

Thanks for finding this!
Please open up a new bug in my repository if you want it fixed. I just mentioned that i created this tool here in case it's useful for others who find themselves needing a decent log viewer, but i think any further discussions on my tool should be done in its repo to not detract from the main problem here, which is that there is no in-built log viewer in rancher.

**Comment by jandubois** [comment]
> Please open up a new bug in my repository if you want it fixed.

I've opened https://github.com/binginsin/LogFollower/issues/1 to let you track it in your own repo.

I can't promise thought that we will add the extension to the built-in catalog; it would also require that it works with `containerd` and not just `moby` (maybe it does already; I didn't test), and we would like to see some other users voice their support for it as well (ideally in a new issue, asking for it to be included).

**Comment by jandubois** [comment]
I think this issue has been implemented by #8811, so I'm going to close this one here.

I'll open a separate issue for the `docker.sock.raw` topic.
